blocks/color-mechanism/index.js

```diff
@@ -0,0 +1,90 @@
+/**
+ * External dependencies
+ */
+import { find, get, kebabCase } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { getWrapperDisplayName } from '@wordpress/element';
+import { withContext } from '@wordpress/components';
+
+/**
+ * Internal dependencies
+ */
+import './style.scss';
+
+const getColorValue = ( colors, namedColor, customColor ) => {
+	if ( namedColor ) {
+		const colorObj = find( colors, { name: namedColor } );
+		return colorObj && colorObj.color;
+	}
+	if ( customColor ) {
+		return customColor;
+	}
+};
+
+const setColorValue = ( colors, colorAttributeName, customColorAttributeName, setAttributes ) =>
+	( colorValue ) => {
+		const colorObj = find( colors, { color: colorValue } );
+		if ( colorObj ) {
+			setAttributes( {
+				[ colorAttributeName ]: colorObj.name,
+				[ customColorAttributeName ]: undefined,
+			} );
+			return;
+		}
+		setAttributes( {
```

Comment:
Seems we could DRY up the `setAttributes` and reduce number of return paths.

```js
setAttributes( {
	[ colorAttributeName ]: colorObj ? colorObj.name : undefined,
	[ customColorAttributeName ]: colorObj ? undefined : colorValue,
} );
```

Or (my preference):

```js
const nextAttributes = {};
if ( colorObj ) {
	nextAttributes[ colorAttributeName ] = colorObj.name;
} else {
	nextAttributes[ customColorAttributeName ] = colorValue;
}

setAttributes( nextAttributes );
```

Reply:
Hi @aduth, I followed the first approach. I did not use it at first to not have to conditions checking for colorObj, but on a second thought, the code looks simples and shorter with this change.
Regarding the second approach, I think it is not equivalent to the first one and would not work. We want to set an attribute and unset the other. The second approach would just set one attribute but leave the other one untouched ( as it was before ).